packages/edit-site/src/components/global-styles/hooks.js

```diff
@@ -0,0 +1,210 @@
+
+/**
+ * Internal dependencies
+ */
+import { store as editSiteStore } from '../../store';
+
+const EMPTY_CONFIG = { isGlobalStylesUserThemeJSON: true, version: 1 };
+
+function useGlobalStylesUserConfig() {
+	const globalStylesId = useSelect( ( select ) => {
+		return select( editSiteStore ).getSettings()
+			.__experimentalGlobalStylesUserEntityId;
+	}, [] );
+
+	const [ content, setContent ] = useEntityProp(
+		'postType',
+		'wp_global_styles',
+		'content',
+		globalStylesId
+	);
+
+	const config = useMemo( () => {
+		let parsedConfig;
+		try {
+			parsedConfig = content ? JSON.parse( content ) : {};
+		} catch ( e ) {
+			/* eslint-disable no-console */
+			console.error( 'Global Styles User data is not valid' );
+			console.error( e );
+			/* eslint-enable no-console */
+			parsedConfig = {};
+		}
+
+		return parsedConfig;
+	}, [ content ] );
+
+	const setConfig = useCallback(
+		( newConfig ) => setContent( JSON.stringify( newConfig ) ),
+		[ setContent ]
+	);
+
+	return [ config, setConfig ];
+}
+
+function useGlobalStylesBaseConfig() {
+	const baseConfig = useSelect( ( select ) => {
+		return select( editSiteStore ).getSettings()
+			.__experimentalGlobalStylesBaseStyles;
+	}, [] );
+
+	return baseConfig;
+}
+
+function useGlobalStylesConfig() {
+	const [ userConfig, setUserConfig ] = useGlobalStylesUserConfig();
+	const baseConfig = useGlobalStylesBaseConfig();
+
+	return [ baseConfig, userConfig, setUserConfig ];
+}
+
+export const useGlobalStylesReset = () => {
+	const [ config, setConfig ] = useGlobalStylesUserConfig();
+	const canReset = !! config && ! isEqual( config, EMPTY_CONFIG );
+	return [
+		canReset,
+		useCallback( () => setConfig( EMPTY_CONFIG ), [ setConfig ] ),
+	];
+};
+
+export function useSetting( path, blockName, source = 'all' ) {
+	const [ baseConfig, userConfig, setUserConfig ] = useGlobalStylesConfig();
+	const finalPath = ! blockName
+		? `settings.${ path }`
+		: `settings.blocks.${ blockName }.${ path }`;
+
+	const getBaseSetting = () => {
+		const result = get( baseConfig, finalPath );
+		if ( PATHS_WITH_MERGE[ path ] ) {
+			return result.theme ?? result.core;
+		}
+	};
+
+	const setSetting = ( newValue ) => {
+		const newUserConfig = cloneDeep( userConfig );
+		set( newUserConfig, finalPath, newValue );
+		setUserConfig( newUserConfig );
+	};
+
+	let result;
+	switch ( source ) {
+		case 'all':
+			result = get( userConfig, finalPath ) ?? getBaseSetting();
+			break;
+		case 'user':
+			result = get( userConfig, finalPath );
+			break;
+		case 'base':
+			result = getBaseSetting();
+			break;
+		default:
+			throw 'Unsupported source';
+	}
+
+	return [ result, setSetting ];
+}
+
+export function useStyle( path, blockName, source = 'all' ) {
+	const [ baseConfig, userConfig, setUserConfig ] = useGlobalStylesConfig();
+	const finalPath = ! blockName
+		? `styles.${ path }`
+		: `styles.blocks.${ blockName }.${ path }`;
+
+	const setStyle = ( newValue ) => {
+		const newUserConfig = cloneDeep( userConfig );
```

Comment:
I guess we could easily improve the performance and even code readability by using the immutableSet function that we already use on other parts of the codebase:
```
function immutableSet( object, path, value ) {
	return setWith( object ? clone( object ) : {}, path, value, clone );
}
```